File: android/app/src/main/java/top/babylink/app/LinkStoragePlugin.java
package top.babylink.app;

import android.os.StatFs;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.File;

@CapacitorPlugin(name = "LinkStorage")
public class LinkStoragePlugin extends Plugin {
    private static final class DirectoryStats {
        long bytes;
        int fileCount;

        void add(DirectoryStats other) {
            bytes += other.bytes;
            fileCount += other.fileCount;
        }
    }

    @PluginMethod
    public void getOverview(PluginCall call) {
        getBridge().execute(() -> {
            try {
                call.resolve(buildOverview());
            } catch (Exception error) {
                call.reject("无法读取 App 存储信息。", error);
            }
        });
    }

    @PluginMethod
    public void clearTemporaryFiles(PluginCall call) {
        getBridge().execute(() -> {
            try {
                clearDirectoryContents(getContext().getCacheDir());
                File externalCacheDirectory = getContext().getExternalCacheDir();
                if (externalCacheDirectory != null) clearDirectoryContents(externalCacheDirectory);
                call.resolve(buildOverview());
            } catch (Exception error) {
                call.reject("清理 App 临时文件失败。", error);
            }
        });
    }

    private JSObject buildOverview() {
        DirectoryStats cacheStats = collectDirectoryStats(getContext().getCacheDir());
        File externalCacheDirectory = getContext().getExternalCacheDir();
        if (externalCacheDirectory != null) cacheStats.add(collectDirectoryStats(externalCacheDirectory));

        StatFs fileSystem = new StatFs(getContext().getFilesDir().getAbsolutePath());
        JSObject result = new JSObject();
        result.put("platform", "android");
        result.put("availableBytes", fileSystem.getAvailableBytes());
        result.put("totalBytes", fileSystem.getTotalBytes());
        result.put("cacheBytes", cacheStats.bytes);
        result.put("cacheFileCount", cacheStats.fileCount);
        return result;
    }

    private DirectoryStats collectDirectoryStats(File directory) {
        DirectoryStats stats = new DirectoryStats();
        if (directory == null || !directory.exists()) return stats;
        File[] children = directory.listFiles();
        if (children == null) return stats;

        for (File child : children) {
            if (child.isDirectory()) {
                stats.add(collectDirectoryStats(child));
            } else if (child.isFile()) {
                stats.bytes += Math.max(0L, child.length());
                stats.fileCount += 1;
            }
        }
        return stats;
    }

    private void clearDirectoryContents(File directory) {
        if (directory == null || !directory.exists()) return;
        File[] children = directory.listFiles();
        if (children == null) return;
        for (File child : children) deleteRecursively(child);
    }

    private void deleteRecursively(File target) {
        if (target.isDirectory()) {
            File[] children = target.listFiles();
            if (children != null) {
                for (File child : children) deleteRecursively(child);
            }
        }
        if (target.exists() && !target.delete()) throw new IllegalStateException("无法删除临时文件：" + target.getName());
    }
}

File: ios/App/App/LinkStoragePlugin.swift
import Capacitor
import Foundation

@objc(LinkStoragePlugin)
final class LinkStoragePlugin: CAPPlugin, CAPBridgedPlugin {
    let identifier = "LinkStoragePlugin"
    let jsName = "LinkStorage"
    let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getOverview", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearTemporaryFiles", returnType: CAPPluginReturnPromise)
    ]

    private struct DirectoryStats {
        var bytes = 0
        var fileCount = 0

        mutating func add(_ other: DirectoryStats) {
            bytes += other.bytes
            fileCount += other.fileCount
        }
    }

    @objc func getOverview(_ call: CAPPluginCall) {
        DispatchQueue.global(qos: .utility).async { [weak self] in
            guard let self else { return }
            call.resolve(self.buildOverview())
        }
    }

    @objc func clearTemporaryFiles(_ call: CAPPluginCall) {
        DispatchQueue.global(qos: .utility).async { [weak self] in
            guard let self else { return }
            do {
                try self.clearDirectoryContents(self.cacheDirectoryURL())
                try self.clearKnownTemporaryBackups()
                call.resolve(self.buildOverview())
            } catch {
                call.reject("清理 App 临时文件失败。", nil, error)
            }
        }
    }

    private func buildOverview() -> [String: Any] {
        var cacheStats = directoryStats(for: cacheDirectoryURL())
        cacheStats.add(directoryStats(for: FileManager.default.temporaryDirectory))

        let storageURL = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? cacheDirectoryURL()
        let values = try? storageURL.resourceValues(forKeys: [
            .volumeAvailableCapacityForImportantUsageKey,
            .volumeTotalCapacityKey
        ])
        return [
            "platform": "ios",
            "availableBytes": values?.volumeAvailableCapacityForImportantUsage ?? 0,
            "totalBytes": values?.volumeTotalCapacity ?? 0,
            "cacheBytes": cacheStats.bytes,
            "cacheFileCount": cacheStats.fileCount
        ]
    }

    private func cacheDirectoryURL() -> URL {
        FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
    }

    private func directoryStats(for directory: URL) -> DirectoryStats {
        let manager = FileManager.default
        guard let enumerator = manager.enumerator(at: directory, includingPropertiesForKeys: [.isRegularFileKey, .fileSizeKey]) else {
            return DirectoryStats()
        }

        var stats = DirectoryStats()
        for case let fileURL as URL in enumerator {
            let values = try? fileURL.resourceValues(forKeys: [.isRegularFileKey, .fileSizeKey])
            guard values?.isRegularFile == true else { continue }
            stats.bytes += values?.fileSize ?? 0
            stats.fileCount += 1
        }
        return stats
    }

    private func clearDirectoryContents(_ directory: URL) throws {
        let manager = FileManager.default
        let children = try manager.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)
        for child in children {
            try manager.removeItem(at: child)
        }
    }

    private func clearKnownTemporaryBackups() throws {
        let directory = FileManager.default.temporaryDirectory
            .appendingPathComponent("BabyLinkBackups", isDirectory: true)
        guard FileManager.default.fileExists(atPath: directory.path) else { return }
        try FileManager.default.removeItem(at: directory)
    }
}